vision/segmentation/er_ccl_clustering.go

```diff
@@ -0,0 +1,274 @@
+	}
+	return cfg.ErCCLAlgorithm, nil
+}
+
+// ErCCLAlgorithm applies the connected components clustering algorithm directly on a given point cloud.
+func (erCCL *ErCCLConfig) ErCCLAlgorithm(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+	// get next point cloud
+	cloud, err := src.NextPointCloud(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// run ransac, get pointcloud without ground plane
+	ps := NewPointCloudGroundPlaneSegmentation(cloud, erCCL.MaxDistFromPlane, erCCL.MinPtsInPlane, erCCL.AngleTolerance, erCCL.NormalVec)
+	// if there are found planes, remove them, and keep all the non-plane points
+	_, nonPlane, err := ps.FindGroundPlane(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// need to figure out coordinate system
+	// if height is not y, then height is going to be z
+	heightIsY := erCCL.NormalVec.Y != 0
+
+	// calculating s value, want 200 x 200 graph
+	resolution := math.Ceil((nonPlane.MetaData().MaxX - nonPlane.MetaData().MinX) / 200)
+	if heightIsY {
+		resolution = math.Ceil((math.Ceil((nonPlane.MetaData().MaxZ-nonPlane.MetaData().MinZ)/200) + resolution) / 2)
+	} else {
+		resolution = math.Ceil((math.Ceil((nonPlane.MetaData().MaxY-nonPlane.MetaData().MinY)/200) + resolution) / 2)
+	}
+
+	// create obstacle flag map, return that 2d slice of nodes
+	labelMap := pcProjection(nonPlane, resolution, heightIsY)
+
+	// actually run erCCLL
+	// iterate through every box, searching down and right r distance
+	// run calculations to meet similarity threshold
+	// if similar enough update to initial label value (will also be smallest)
+	// iterate through pointcloud
+
+	i := 0
+	continueRunning := true
+	for continueRunning {
+		// 0.9 is alpha
+		continueRunning := labelMapUpdate(labelMap, erCCL.ClusteringRadius, 0.9, erCCL.Beta, resolution)
+		if !continueRunning {
+			break
+		}
+
+		if i > 300000 { // arbitrary cutoff for iterations
+			return nil, errors.New("could not converge, change parameters")
+		}
+		i++
+	}
+
+	// look up label value of point by looking at 2d array and seeing what label inside that struct
+	// set this label
+	segments := NewSegments()
+	nonPlane.Iterate(0, 0, func(p r3.Vector, d pc.Data) bool {
+		i := int(math.Ceil((p.X - nonPlane.MetaData().MinX) / resolution))
+		j := int(math.Ceil((p.Z - nonPlane.MetaData().MinZ) / resolution))
+		if !heightIsY {
+			j = int(math.Ceil((p.Y - nonPlane.MetaData().MinY) / resolution))
+		}
+		// fmt.Println("i:", i, ", j:", j)
+		err := segments.AssignCluster(p, d, labelMap[i][j].label)
+		if err != nil {
+			panic("clustering went wrong uhhhh")
+		}
+		return true
+	})
+	// prune smaller clusters
+	validClouds := pc.PrunePointClouds(segments.PointClouds(), erCCL.MinPtsInSegment)
+	// wrap
+	objects, err := NewSegmentsFromSlice(validClouds, "")
+	if err != nil {
+		return nil, err
+	}
+	return objects.Objects, nil
+	// this seems a bit wasteful to make segments then make more segments after filtering, but rolling with it for now
```

Comment:
create an improvements ticket, and assign this as a TODO with the ticket number in the comment